### PR TITLE
Merge AMPOptimizerCallback and OptimizerCallback

### PR DESCRIPTION
**Pull request by and-kul**
## Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [ ] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [x] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [ ] Did you check the docs with `make check-docs`?
- [ ] Did you write any new necessary tests?
- [x] Did you check that your code passes the unit tests `pytest .` ? 
- [x] Did you add your new functionality to the docs?
- [ ] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md)?

## Description

* Remove `AMPOptimizerCallback` and move its functionality into `OptimizerCallback`
* Change logic of `fp16=True` in `runner.train`:
```
If fp16=True, params by default will be:
    * ``{"amp": True}`` if torch>=1.6.0
    * ``{"apex": True, "opt_level": "O1", ...}`` if torch<1.6.0
```
* `OptimizerCallback` now have 2 new args: `use_amp` and `use_apex`. They can be set manually or inferred from `runner.experiment.distributed_params`
* `decouple_weight_decay` by default is now `False`
    * This change is discussible, but it seems like this is a safer assumption (for example we now have `AdamW` optimizer in PyTorch itself, which implements decoupling logic)

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

PS
- [x] I know, that I could [join slack](https://join.slack.com/t/catalyst-team-core/shared_invite/zt-d9miirnn-z86oKDzFMKlMG4fgFdZafw) for pull request discussion.

### CODE REVIEW

File: catalyst/runners/runner.py [hunk 1]
```diff
@@ -30,6 +31,33 @@
 from catalyst.utils.tracing import save_traced_model, trace_model
 
 
+def resolve_bool_fp16(fp16: Union[Dict, bool]):
```
Comment (Scitator) [inline catalyst/runners/runner.py]: could we make it `_resolve_bool_fp16`? or move to `catalyst/runner/functional.py`? I just want to keep `runner.py` clean and clear :)

File: catalyst/runners/runner.py [hunk 2]
```diff
@@ -397,14 +425,10 @@ def trace(
...
-        elif isinstance(fp16, bool) and not fp16:
-            opt_level = None
-        elif isinstance(fp16, dict):
-            opt_level = fp16["opt_level"]
-        else:
-            opt_level = fp16
+        fp16 = resolve_bool_fp16(fp16)
+        opt_level = None
+        if fp16:
+            fp16.get("opt_level", None)
```
Comment (Scitator) [inline catalyst/runners/runner.py]:
```suggestion
            opt_level = fp16.get("opt_level", None)
```

Comment (and-kul) [reply catalyst/runners/runner.py]: Точно, спасибо

File: catalyst/runners/runner.py (hunk 1 above)
Comment (and-kul) [reply catalyst/runners/runner.py]: Done

**Pull request merged**